tokens/usecase/pricing/coingecko/pricing_coingecko.go

```diff
@@ -0,0 +1,168 @@
+)
+
+var (
+	cacheHitsCounter = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "sqs_pricing_coingecko_cache_hits_total",
+			Help: "Total number of pricing coingecko cache hits",
+		},
+		[]string{"base", "quote"},
+	)
+	cacheMissesCounter = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "sqs_pricing_coingecko_cache_misses_total",
+			Help: "Total number of pricing coingecko cache misses",
+		},
+		[]string{"base", "quote"},
+	)
+)
+
+// CoingeckoPriceGetterFn is a function type that fetches the price of a token from Coingecko.
+// We monkey-patch this function for testing purposes.
+type CoingeckoPriceGetterFn func(ctx context.Context, baseDenom string, coingeckoId string) (osmomath.BigDec, error)
+
+// DefaultCoingeckoPriceGetter is the default implementation of CoingeckoPriceGetterFn, which invokes GetPriceByCoingeckoId.
+var DefaultCoingeckoPriceGetter CoingeckoPriceGetterFn = nil
+
+// MockCoingeckoPriceGetter is a mock implementation of CoingeckoPriceGetterFn, which returns a fixed price of 1.
+var MockCoingeckoPriceGetter CoingeckoPriceGetterFn = func(ctx context.Context, baseDenom, coingeckoId string) (osmomath.BigDec, error) {
+	if coingeckoId == "" {
+		return osmomath.NewBigDec(0), nil
+	} else {
+		return osmomath.NewBigDec(1), nil
+	}
+}
+
+type coingeckoPricing struct {
+	TUsecase      mvc.TokensUsecase
+	cache         *cache.Cache
+	cacheExpiryNs time.Duration
+	quoteCurrency string
+	coingeckoUrl  string
+
+	// We monkey-patch this function for testing purposes.
+	priceGetterFn CoingeckoPriceGetterFn
+}
+
+func init() {
+	if err := prometheus.Register(cacheHitsCounter); err != nil {
+		panic(err)
+	}
+	if err := prometheus.Register(cacheMissesCounter); err != nil {
+		panic(err)
+	}
+}
+
+// New creates a new Coingecko pricing source.
+// if coinGeckoPriceGetterFn is nil, it uses the default implementation.
+func New(routerUseCase mvc.RouterUsecase, tokenUseCase mvc.TokensUsecase, config domain.PricingConfig, coingeckoPriceGetterFn CoingeckoPriceGetterFn) domain.PricingSource {
+	coingeckoPricing := &coingeckoPricing{
+		TUsecase:      tokenUseCase,
+		cache:         cache.New(),
+		cacheExpiryNs: time.Duration(config.CacheExpiryMs) * time.Millisecond,
+		quoteCurrency: config.CoingeckoQuoteCurrency,
+		coingeckoUrl:  config.CoingeckoUrl,
+	}
+
+	if coingeckoPriceGetterFn == nil {
+		// Set the default price getter function.
+		coingeckoPricing.priceGetterFn = coingeckoPricing.GetPriceByCoingeckoId
+	} else {
+		// Set the custom price getter function (useful for testing purposes)
+		coingeckoPricing.priceGetterFn = coingeckoPriceGetterFn
+	}
+
+	return coingeckoPricing
+}
+
+// GetPrice implements pricing.PricingStrategy.
+// quoteDenom is ignored as it uses always coingecko-quote-currency in config.json
+func (c *coingeckoPricing) GetPrice(ctx context.Context, baseDenom string, quoteDenom string, opts ...domain.PricingOption) (osmomath.BigDec, error) {
```

Comment:
Using the mocking infrastructure that was just integrated, would it be possible to add more table-driven unit-tests that maximize all of the edge case coverage while treating this function as a black box?

For example, covering the fact that `quoteDenom` is ignored via a table-driven unit test case.

I know that we already have a higher-level test in `tokensUseCase` but I think that also having a stricter test at a lower level of abstraction would pay off in the long term

Reply:
Having table-driven unit test cases make sense. But as you said, there may be some test cases duplication with `tokens_usecase_test.go`, what edge cases are you you thinking about? 

Thanks!

Reply:
It is acceptable to have overlapping unit tests at various levels of abstraction.

Having them defined in such a way would allow us, developers, to identify the exact source of the bug or a regression if any arise.

Ideally, we should aim to treat a function as a black box. Given the abstraction, its state, and all possible combinations of inputs and outputs - what are the test cases that make sense to construct? 

In this case, I would add the following cases:
- happy path case (quote denom defined on the struct is used and same quote denom is provided)
- the quote denom provided as input to the function is ignored and another denom that is defined on the struct
- some other quote denom is set on the strcut (e.g. instead of usdc, set usdt) 
- mock error being returned and bubbled up to the client
  * This case is important because we have various assumptions around the system around error handling. In some areas, we silently ignore them to avoid breaking the client and return zero instead. In other cases, we actually bubble up the error. As a result, it is important to keep these invariants strict with test coverage.

Reply:
True - it does help identifying where the root cause may be. Let's sync up tomorrow (Wed). I'll show you what I've done so far and make sure I'm on the right track. Thanks!

Reply:
@p0mvn , I've enhanced and test cases. Feel free to comment it. Thanks!!